compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Upgrade.hs

```diff
@@ -108,124 +126,114 @@ checkModule module_ = do
-        allChoiceReturnTCons module_ = HMS.fromList $ do
-            template <- NM.toList (moduleTemplates module_)
-            choice <- NM.toList (tplChoices template)
-            dtName <- flip cata (chcReturnType choice) $ \case
-                TConF dtName -> [dtName]
-                rest -> fold rest
-            pure (qualObject dtName, (template, choice))
+        deriveVariantInfo :: LF.Module -> HMS.HashMap LF.TypeConName (LF.DefDataType, LF.VariantConName)
+        deriveVariantInfo module_ = HMS.fromList $ do
+            dataType <- NM.toList (moduleDataTypes module_)
+            DataVariant variants <- pure $ dataCons dataType
+            (variantName, TConApp recordName _) <- variants
+            pure (qualObject recordName, (dataType, variantName))
         dataTypeOrigin
             :: DefDataType -> Module
-            -> (UpgradedRecordOrigin, (Context, Bool))
+            -> (UpgradedRecordOrigin, Context)
         dataTypeOrigin dt module_
             | Just template <- NM.name dt `NM.lookup` moduleTemplates module_ =
                 ( TemplateBody (NM.name dt)
-                , ( ContextTemplate module_ template TPWhole
-                  , True
-                  )
+                , ContextTemplate module_ template TPWhole
                 )
             | Just (template, choice) <- NM.name dt `HMS.lookup` deriveChoiceInfo module_ =
                 ( TemplateChoiceInput (NM.name template) (NM.name choice)
-                , ( ContextTemplate module_ template (TPChoice choice)
-                  , True
-                  )
+                , ContextTemplate module_ template (TPChoice choice)
+                )
+            | Just (variant, variantName) <- NM.name dt `HMS.lookup` deriveVariantInfo module_ =
+                ( VariantConstructor (dataTypeCon variant) variantName
+                , ContextDefDataType module_ variant
                 )
             | otherwise =
-                ( TopLevel
-                , (ContextDefDataType module_ dt
-                  , NM.name dt `HMS.member` allChoiceReturnTCons module_
-                  )
+                ( TopLevel (dataTypeCon dt)
+                , ContextDefDataType module_ dt
                 )
 
     forM_ dtExisting $ \dt ->
-        -- Get origin/context for each datatype in both past and present
+        -- Get origin/context for each datatype in both _past and _present
         let origin = dataTypeOrigin <$> dt <*> module_
         in
         -- If origins don't match, record has changed origin
         if foldU (/=) (fst <$> origin) then
-            withContext (ContextDefDataType (present module_) (present dt)) $
-                throwWithContext (EUpgradeError (RecordChangedOrigin (dataTypeCon (present dt)) (fst (past origin)) (fst (present origin))))
-        else
-            let (presentOrigin, (context, shouldCheck)) = present origin
-            in
-            when shouldCheck $
```

Comment:
You should always be able to add new data types or templates in a upgrade compatible way (unless it's a name clash).

Ideally you would allow internal data types to be changed. Problem is we don't define explicitly what internal is, so indeed we would need to check that the data type is not exported or possibly put on the ledger as part of a template. This is not great UX either way. 

I would keep it simple for 2.9, and disallow all, including "internal" data types to change in non-upgradeable ways. There have been requests to support a private keyword, possibly for Daml 3. At that point it seems reasonable to support making this check more permissible.

Reply:
Yeah, now we're upgrade-checking all of the datatypes in a module.